WordPress/gutenberg
# Memoize getEntityRecords to prevent infinite re-renders

## Description


Memoize `reusableBlocks` to prevent infinite re-renders in some cases.

`reusableBlocks` is called from `getEntityRecords` here:
https://github.com/WordPress/gutenberg/blob/a4af7d0a82c8dd16e4f72b23128aef3646cf8f69/packages/edit-widgets/src/components/widget-areas-block-editor-provider/index.js#L45-L48

When the data hasn't arrived yet, `getEntityRecords` will return an empty array as defined here:
https://github.com/WordPress/gutenberg/blob/a4af7d0a82c8dd16e4f72b23128aef3646cf8f69/packages/core-data/src/selectors.js#L240-L242

Keep in mind that, for every change to the store, `getEntityRecords` will return a new instance of the empty array.

`reusableBlocks` is then passed to `useMemo` below to form a `settings` object.
https://github.com/WordPress/gutenberg/blob/a4af7d0a82c8dd16e4f72b23128aef3646cf8f69/packages/edit-widgets/src/components/widget-areas-block-editor-provider/index.js#L74

`settings` is passed to `<BlockEditorProvider>` below, which schedules an effect to update the settings whenever it changes:
https://github.com/WordPress/gutenberg/blob/a4af7d0a82c8dd16e4f72b23128aef3646cf8f69/packages/block-editor/src/components/provider/index.js#L19-L21

This process creates an infinite loop before the `quriedData` arrived.

1. `getEntityRecords` returns a new empty array, hence a new `reusableBlocks`.
2. Since `reusableBlocks` changes, `useMemo` computes a new `settings`.
3. Since `settings` changes, `useEffect` calls `updateSettings` to update the store.
4. Since the store changes, `useSelect` calls `getEntityRecords` again and returns a new empty array, back to (1).

This loop only ends when `getEntityRecords` returns the `queriedData` returned from the network request in (1).

This PR tries to fix this by memoize `reusableBlocks` so that we don't have to create a new `settings` whenever `getEntityRecords` returns a new array, breaking the loop in (2).

## How has this been tested?



1. Add an arbitrary `console.log` in the `useMemo` of `settings`
2. Reload the page
3. Observe that the log gets called several times before the network response arrived
4. Apply this PR, and reload the page
5. Observe that there's only 3 calls now.

See the screenshots below.

## Screenshots 

(It really shouldn't be `rendered` in the log but hopefully you get the point 😅 )

Before | After
--- | ---
![Screen Shot 2020-10-26 at 5 18 22 PM](https://user-images.githubusercontent.com/7753001/97156476-b77ced00-17b1-11eb-96d4-34e7972b9b9b.png) | ![Screen Shot 2020-10-26 at 5 17 32 PM](https://user-images.githubusercontent.com/7753001/97156467-b3e96600-17b1-11eb-9da3-200a534a3cec.png)

## Types of changes




Bug fix

## Checklist:
